rosbag2_cpp/src/rosbag2_cpp/cache/message_cache.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rosbag2_cpp/cache/message_cache.hpp"
+#include "rosbag2_cpp/logging.hpp"
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+MessageCache::MessageCache(const uint64_t & max_buffer_size)
+{
+  primary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+  secondary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+}
+
+bool MessageCache::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  // While pushing, we keep track of inserted and dropped messages as well
+  bool pushed = false;
+  {
+    std::lock_guard<std::mutex> buffer_lock(buffer_mutex_);
+    pushed = primary_buffer_->push(msg);
+  }
+
+  allow_swap();
+
+  if (!pushed) {
+    messages_dropped_per_topic_[msg->topic_name]++;
+  }
+  return pushed;
+}
+
+void MessageCache::allow_swap()
```

Comment:
sorry, don't want to be picky here, but we might as well rename this function to `notify_swap` to be on pair with `wait_for_swap`. 
Even though in general, I am still not too happy about the expression "swap" as I don't think it indicates fully what's going to happen. Does "wait_for_buffer` and `notify_buffer_swap` make sense?

Reply:
I am opting for wait_for_buffer() and notify_buffer_consumer(), would that do?

Reply:
sounds good.